prover/src/full/prove.rs

```diff
@@ -0,0 +1,206 @@
+use crate::{
+    default_values::{
+        KECCAK_GATE_CONFIG, KECCAK_PK, KECCAK_PROOF, KECCAK_PROTOCOL,
+        KECCAK_SRS, OUTER_GATE_CONFIG, OUTER_PK, OUTER_PROOF, OUTER_SRS,
+        UBV_GATE_CONFIG, UBV_PK, UBV_PROOF, UBV_PROTOCOL, UBV_SRS, UPA_CONFIG,
+    },
+    keccak, universal_batch_verifier, universal_outer,
+};
+use circuits::utils::upa_config::UpaConfig;
+use clap::Parser;
+use log::info;
+
+#[derive(Clone, Debug, Parser)]
+pub struct ProveParams {
+    /// UPA Config file
+    #[arg(short = 'c', long, value_name = "config-file", default_value = UPA_CONFIG)]
+    config: String,
+
+    /// UBV circuit SRS file
+    #[arg(long, value_name = "ubv-srs-file", default_value = UBV_SRS)]
+    ubv_srs: String,
+
+    /// Keccak circuit SRS file
+    #[arg(long, value_name = "keccak-srs-file", default_value = KECCAK_SRS)]
+    keccak_srs: String,
+
+    /// Outer circuit SRS file
+    #[arg(long, value_name = "outer-srs-file", default_value = OUTER_SRS)]
+    outer_srs: String,
+
+    /// UBV proving key file
+    #[arg(long, value_name = "ubv-proving-key-file", default_value = UBV_PK)]
+    ubv_proving_key: String,
+
+    /// Keccak proving key file
+    #[arg(long, value_name = "keccak-proving-key-file", default_value = KECCAK_PK)]
+    keccak_proving_key: String,
+
+    /// Outer proving key file
+    #[arg(long, value_name = "outer-proving-key-file", default_value = OUTER_PK)]
+    outer_proving_key: String,
+
+    /// Circuit specs file (UBVGateConfig)
+    #[arg(long, value_name = "ubv-gate-config-file", default_value = UBV_GATE_CONFIG)]
+    ubv_gate_config: String,
+
+    /// Circuit specs file (KeccakGateConfig)
+    #[arg(long, value_name = "keccak-gate-config-file", default_value = KECCAK_GATE_CONFIG)]
+    keccak_gate_config: String,
+
+    /// Circuit specs file (OuterGateConfig)
+    #[arg(long, value_name = "outer-gate-config-file", default_value = OUTER_GATE_CONFIG)]
+    outer_gate_config: String,
+
+    /// Universal Batch Verify circuit protocol file
+    #[arg(long, value_name = "bv-circuit-protocol-file", default_value = UBV_PROTOCOL)]
+    ubv_protocol: String,
+
+    /// Keccak circuit protocol file
+    #[arg(long, value_name = "keccak-protocol-file", default_value = KECCAK_PROTOCOL)]
+    keccak_protocol: String,
+
+    /// JSON files each containing a batch of app_vk, proof, public input triples
+    /// to be verified by the Universal Batch Verifier circuit.
+    #[arg(long, value_name = "app-vk-proof-batch-file")]
+    app_vk_proof_batch: Vec<String>,
+
+    /// Output UBV proof file
+    #[arg(long, value_name = "ubv-proof-file", default_value = UBV_PROOF)]
```

Comment:
This seems more like the base of a sequence of file names, rather an actual file name.

Reply:
it could be a sequence, I agree. But it you look at the code below we're naming them with `ubv_proof_name_{i}` for `i` the outer index.

Reply:
```suggestion
    #[arg(long, value_name = "ubv-proof-file-base", default_value = UBV_PROOF_BASE)]
```

Reply:
> But it you look at the code below we're naming them with ubv_proof_name_{i} for i the outer index.

That's my point.  If the called writes `--ubv-proof-file somefile` they expect a file `somefile` to be generated.  But that's not what's happening.

Reply:
So the use of the basename is totally fine and makes sense, I'm just suggesting we call it that in the external interface.

Reply:
changed